omegaconf/dictconfig.py

```diff
@@ -176,73 +192,97 @@ def _validate_set_merge_impl(self, key: Any, value: Any, is_assign: bool) -> Non
         )
 
-        if not is_assign:
-            # merge
-            # Merging of a dictionary is allowed even if assignment is illegal (merge would do deeper checks)
-            validation_error = not is_dict(value_type) and validation_error
-
         if validation_error:
-            assert value_type is not None
-            assert target_type is not None
-            msg = (
-                f"Invalid type assigned : {type_str(value_type)} is not a "
-                f"subclass of {type_str(target_type)}. value: {value}"
-            )
-            raise ValidationError(msg)
+            self._raise_invalid_value(value, value_type, target_type)
 
-    def _check_assign_value_node(self, key: Any, value: Any) -> None:
-        from omegaconf import OmegaConf
+    def _raise_invalid_value(
+        self, value: Any, value_type: Any, target_type: Any
+    ) -> None:
+        assert value_type is not None
+        assert target_type is not None
+        msg = (
+            f"Invalid type assigned : {type_str(value_type)} is not a "
+            f"subclass of {type_str(target_type)}. value: {value}"
+        )
+        raise ValidationError(msg)
 
-        element_type = self._metadata.element_type
-        value_type = OmegaConf.get_type(value)
-        if value_type is not Any and not issubclass(value_type, element_type):  # type: ignore
-            msg = (
-                f"Invalid type assigned : {type_str(value_type)} is not a "
-                f"subclass of {type_str(element_type)}. value: {value}"
-            )
-            raise ValidationError(msg)
+    def _is_special_value(self, value: Any) -> bool:
+        # target must be optional by now. no need to check the type of value if None.
+        return value == "???" or value is None
+
+    def _get_assign_target_type(self, target: Optional[Node], value: Any) -> Any:
+        if self._is_value_node(value):
+            target_type = self._metadata.element_type
+        else:
+            target_type = target._metadata.ref_type  # type: ignore
+        return target_type
+
+    def _is_invalid_target_value_type(self, value_type: Any, target_type: Any) -> bool:
+        invalid = False
+        if is_generic_type(target_type):
+            invalid = True
+        elif is_dict(value_type) and is_dict(target_type):
+            invalid = True
+        return invalid
+
+    def _is_valid_target(self, target: Optional[Node]) -> bool:
+        valid = True
+        if target is None or not is_typed_dictconfig(target):
+            valid = False
+        return valid
+
+    def _is_value_node(self, value: Any) -> bool:
+        valid = False
+        if isinstance(value, ValueNode) and self._has_element_type():
+            valid = True
+        return valid
+
+    def _get_target(self, key: Any) -> Optional[Node]:
+        target: Optional[Node]
+        if key is None:
+            target = self
+        else:
+            target = self._get_node(key)
+        return target
```

Comment:
I don't like this.
you added a bunch of tiny functions that are sensible only in when called from the the validation and does not make any sense in any other context.

I would rather have a minimal amount of duplication between the validate_set and validate_merge than to add a page of code that is not useful in any other context.

Reply:
I understand what you mean but, wouldn't that make it less readable ?

Reply:
I may seem that adding many small functions enhance readability but it's not always a clear cut.
If I have to jump around the code a lot to understand why it's doing it's reducing readability.
If something is just being used once, I would rather have the code where it's being used.
You can use local boolean variables to enhance readability.
You can even use local functions if you feel something is best described in a function. This way you do not contaminate the module with utility functions that are not a part of the API.